contents/docs/self-host/runbook/plugin-server/wtf.mdx

```diff
@@ -0,0 +1,289 @@
+
+All the services listed here run concurrently. 
+
+##### Kafka Consumer
+
+The Kafka consumer consumes from a specified Kafka topic and triggers a run of the event pipeline in the worker thread. The pipeline may be run partially or in full, and all that it does will depend on the mode the server is running in.
+
+##### PubSub
+
+The PubSub service connects to [Redis Pub/Sub channel](https://redis.io/docs/manual/pubsub/) and is used to trigger plugin reloads (e.g. if the code or config has changed) as well as action reloads (e.g. if an action definition has changed).
+
+##### Graphile Queue
+
+We use the [Graphile Worker](https://github.com/graphile/worker) for our plugin jobs implementation. This queue is a consumer that will periodically check the jobs listed in Postgres and trigger job runs when the jobs target time is reached.
+
+
+##### Schedule
+
+A [node-schedule](https://github.com/node-schedule/node-schedule) scheduler that will trigger scheduled plugin tasks (`runEveryMinute`, `runEveryDay`, etc). 
+
+However, tasks are not necessarily triggered in a given server when the scheduled time hits. We use [Redlock](https://redis.com/redis-best-practices/communication-patterns/redlock/) to ensure that only one plugin server across the fleet processes scheduled tasks at a given point in time.
+
+##### MMDB
+
+The MMDB service uses a [MaxMind](https://www.maxmind.com/en/home) database we download from mmdb.posthog.net (managed by [this service](https://github.com/PostHog/http-mmdb)) and keep in memory in the main thread to allow for geolocation data lookups by IP.
+
+The worker thread communicates with the main thread via TCP to fetch the necessary data.
+
+##### HTTP Server
+
+A simple HTTP server that's currently only used for healthchecks.
+
+
+
+#### Communication between the main thread and worker threads
+
+We use the [Piscina thread pool](https://github.com/piscinajs/piscina) to communicate between the main thread and the worker threads in the plugin server.
+
+By default, Piscina will spawn as many worker threads as the machine's CPU cores, and each worker can process 10 tasks concurrently by default. 
+
+When a task is triggered, Piscina will select an available worker to process it. We've also forked Piscina and added a feature that allows us to run a task on **all** workers. We use this when we need to reload a plugin that has changed, for instance.
+
+#### Worker thread
+
+Most often, a task triggered from the main thread will either immediately trigger a plugin function, or run the event pipeline.
+
+The former happens in the case of "standalone" functions, like scheduled tasks and jobs. The task payload contains everything the VM needs to execute the function, so we call it immediately.
+
+On the other hand, the event pipeline handles the entire lifecycle of an event. The steps in the pipeline are the following:
```

Comment:
```suggestion
On the other hand, the event pipeline handles the entire lifecycle of an event. The steps currently in the pipeline are the following:
```

Note that this may change relatively rapidly (hopefully even next sprint) - but describing current state is still useful!

Reply:
Yeah I was worried about getting too deep into the pipeline (we discussed on the call how I was being a bit generic "process event" / "ingest event") but yeah this is a good change.

Also I probably should clarify that the buffer doesn't do anything yet